src/OrbitCaptureClient/include/OrbitCaptureClient/ApiEventProcessor.h

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_CAPTURE_CLIENT_API_EVENT_PROCESSOR_H_
+#define ORBIT_CAPTURE_CLIENT_API_EVENT_PROCESSOR_H_
+
+#include <absl/container/flat_hash_map.h>
+
+#include "Api/EncodedEvent.h"
+#include "OrbitCaptureClient/CaptureListener.h"
+#include "capture.pb.h"
+
+class ApiEventProcessor {
+ public:
+  ApiEventProcessor(CaptureListener* listener);
+  void ProcessApiEvent(const orbit_grpc_protos::ApiEvent& event_buffer);
+
+ private:
+  void ProcessApiEvent(const orbit_api::ApiEvent& api_event);
+  void ProcessStartEvent(const orbit_api::ApiEvent& api_event);
+  void ProcessStopEvent(const orbit_api::ApiEvent& api_event);
+  void ProcessAsyncStartEvent(const orbit_api::ApiEvent& api_event);
+  void ProcessAsyncStopEvent(const orbit_api::ApiEvent& api_event);
+  void ProcessTrackingEvent(const orbit_api::ApiEvent& api_event);
+
+ private:
+  CaptureListener* capture_listener_ = nullptr;
```

Comment:
Have you considered aligning this class to the model of `GpuQueueSubmissionProcessor` and returning `TimerInfo`s instead of passing the `CaptureListener` to this class? It's weird to me that `GpuQueueSubmissionProcessor` and this class follow opposite models when there's a clear parallel between the two.

Reply:
Not all calls to ProcessApiEvent could return a TimerInfo, "start" events are cached until a corresponding "stop" event arrives, only then can a TimerInfo be created and passed to a listener.

Reply:
You could just return `std::optional<TimerInfo>`.

Reply:
I like that ApiEventProccessor can work both in isolation and as part of a CaptureEventProcessor (this is used in ApiEventProcessorTest.cpp). I prefer keeping it as is instead of introducing std::optional which I think could be confusing. I would find it easier as a reader to look at code sites where the listener is used rather than trying to figure out why the ApiEventProcessor sometimes returns a TimerInfo and sometimes doesn't.

Reply:
> I like that ApiEventProccessor can work both in isolation and as part of a CaptureEventProcessor

Returning `std::optional<TimerInfo>` increases isolation because it removes the dependency on the listener, so I don't understand this point. 

> introducing std::optional which I think could be confusing. I would find it easier as a reader to look at code sites where the listener is used rather than trying to figure out why the ApiEventProcessor sometimes returns a TimerInfo and sometimes doesn't.

I'm not sure about this. Either way, the inconsistency with `GpuQueueSubmissionProcessor` is more confusing. For sure as a reader I'm wondering why one class uses one approach and the other a different one.